File: 0240-search-a-2d-matrix-ii/0240-search-a-2d-matrix-ii.cpp
// SC - O(1)
// TC- O(m+n)
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int rows = matrix.size();
        int cols = matrix[0].size();

        // taking top right corner. so if target is less, you move left. else you move down.
        int r = 0;
        int c = cols-1;

        while(r>=0 and r<rows and c>=0 and c<cols){
            if(matrix[r][c]==target)
                return true;
            else if(matrix[r][c]<target)
                r++;
            else if(matrix[r][c]>target)
                c--;
        }
        return false;
    }
};

File: 0240-search-a-2d-matrix-ii/0240-search-a-2d-matrix-ii.py
# TC- O(m+n)
# SC- O(1)

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        rows, cols = len(matrix), len(matrix[0])

        r, c = 0, cols-1
        while(r<rows and r>=0 and c<cols and c>=0):
            if target==matrix[r][c]:
                return True
            elif target<matrix[r][c]:
                c = c-1
            elif target>matrix[r][c]:
                r = r+1
        return False